tools/ExampleExtractor/ExtractorAndTesterUsersGuide.md

```diff
@@ -0,0 +1,483 @@
+
+In the following example, the class-library source-code is written to file “NestedClassDependency”:
+
+````
+> <!-- Example: {template:"standalone-lib", name:"NestedClassDependency"} -->
+> ```csharp
+> class A
+> {
+>     class B : A {}
+> }
+> ```
+````
+
+### Ellipsis Processing
+
+The treatment of ellipses is controlled by the *replace_ellipsis* annotation directive, which is optional.
+
+```ANTLR
+replace_ellipsis
+    : 'replaceEllipsis' ':' ('true' | 'false')
+    ;
+```
+The C# spec contains many examples having one or more occurrences of an ellipsis (`…`) as a placeholder for content not relevant to the discussion. For example:
+
+````
+> <!-- Example: {template:"standalone-console", name:"ReplaceEllipsisTrue", replaceEllipsis:true}    -->
+> ```csharp
+> class B<U,V> {...}
+> class G<T> : B<string,T[]> {...}
+> ```
+````
+
+When `true` is specified, each ellipsis—regardless of context (including inside string literals)—is replaced with `/* ... */`. When `false` is specified, or the annotation directive is omitted, ellipses are preserved, as written.
+
+Specifying this *annotation_directive* when the example contains no ellipses, has no ill-effect.
+
+> TODO: Unfortunately, such ellipsis replacement does *not* always result in syntactically correct code. When an ellipsis is used as the (unstated) body of a non-void method or a get accessor, this results in compilation error CS0161, “not all code paths return a value.” For example:
+>
+> ```csharp
+> class Gen<T,U>
+> {
+>     public T[,] a;
+>     public void G(int i, T t, Gen<U,T> gt) {/* ... */}  // OK, void method
+>     public U Prop { get {/* ... */} set {/* ... */} }   // CS0161, getter; setter OK
+>     public int H(double d) {/* ... */}                  // CS0161, non-void method
+> }
+> ```
+
+### Expected Compiler Errors
+
+Some tests are expected to fail with one or more compilation errors, in which case, the list of expected error numbers is provided by the *expected_errors* annotation directive, so the numbers in that list can be verified. This *annotation_directive* is optional.
+
+```ANTLR
+expected_errors
+    : 'expectedErrors' ':' '[' cs_number (',' cs_number)* ']'
+    ;
+cs_number
+    : JSON_string_value      // of the form "CSnnnn"
+    ;
+```
+
+The `nnnn` in `"CSnnnn"` is a 4-digit number (with leading zeros, if necessary) [as used by Microsoft’s compiler](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/) for the error message expected.
+
+Error *cs_number*s shall appear in the order in which they are output to the console during compilation. When the compiler produces both error and warning messages, they may be interspersed; however, that has no impact on the checking order of this annotation directive’s list. 
```

Comment:
It turns out that the reporting order isn't deterministic, so I'm changing it to be ordered by line number then error ID.

Reply:
(That requires changing a few examples, but that's okay.)

Reply:
@jskeet I'm not quite sure how I should change the words here.